HGCAL digitisation: end-of-life updates, more realism for startup

franzoni opened the pull request:
#### PR description:

+ improvements and extensions to handle of HGCAL end-of-life scenario
. improvements to the ageing mode (tools to toggle on/off different ingredients)
. fix for the treatment of the common noise, set minimum charge collection efficiency to 0

+ tool to apply realistic noise/gain setting also to startup scenarios ( `SLHCUpgradeSimulations/Configuration/python/aging.py:realisticHGCalStartup` )

+ easier to handle also in startup scenarios (reshuffling `HGCalSiNoiseMap::getSiOpCharacteristics`)
   in this PR the startup scenario is not touched, only tools prepared for the next PR

+ HGCSample:
. bits allocation changed to 12 bits reserved for data Data ( see HGC TDR page 161: adc mode only needs 10, ToT instead needs 12); 4 newly reserved for gain (currently only 3 gains used; 12 are technically possible and may be used)
. boolean `toaFired_` removed, one bit reserve to it


#### PR validation:

Occupancies (only pileup, average 200) :

![radialprofile_counts_CEE_5_ps_gf](https://user-images.githubusercontent.com/4729828/68221990-97757c00-ffea-11e9-81e2-2e00829aaa13.png)


at end-of-life with these improvements have been validated against a simple analytical model which, starting form the sole occupancy of in-time hits w/o noise, predicts the extra occupancy stemming from noise and from out of time pileup. Comparing pred and FS (full-sim) in the last four columns of the table supports the validity of the occupancies form this PR

<img width="863" alt="Screenshot 2019-11-05 at 16 32 16" src="https://user-images.githubusercontent.com/4729828/68221563-e5d64b00-ffe9-11e9-867a-d16a46611e4d.png">


These updated occupancies and the validation model have  been presented at the HGCAL DPG see [here](https://indico.cern.ch/event/855542/contributions/3602119/).

#### if this PR is a backport please specify the original PR:

This PR is not a back-port

@pfs @deguio

Commit:
fix setter of HGCSample

Commit:
add test for HGCSample

Commit:
mega verbosity, eventually to be reverted

Commit:
2 bits used HGCSample to encode the gainID {0,1,2}

Commit:
HGCalSiNoiseMap: remove maxADC, legacy (unused) variable

Commit:
make use of gain info (2 bits) in  HGCSample i.e. propagate gain to cHGCFEElectronics
(conf resolved)

pfs committed:
moving cns to affect only baseline electronics noise

pfs committed:
fixing flooring as well

Commit:
set minimum charge collection efficiency to 0. (prevent polinomials to go negative in rare cases)

pfs committed:
adding algo integer to cfg in order to switch on/off different contributions in the radiation map tool

pfs committed:
moving algo specification to a bitset, propagating changes to dependent classes

pfs committed:
adding algo to aging.py

pfs committed:
fixing some type inconsistencies for scaleByDoseAlgo

pfs committed:
Update HGCalSiNoiseMap.cc

Commit:
Merged add_HGCSample_gain_fix_cns_CMSSW_11_0_0_pre10 from repository PFCal-dev with cms-merge-topic

Commit:
number of bits for ADC set to 10 (was 12 - unnecessary)

Commit:
fix the bit-wise shifts (shift was 1 bit too many for adc)

Commit:
use setters instsad of repeating the bit-manipulation code (time performance improved)

Commit:
remove toaFired_ as data member, use one of the bits of value_ instead

Commit:
introduce getWord to avoid code repetition in getters

Commit:
readibilit

Commit:
extend and cleanup test program

Commit:
blanks

Commit:
12 bits reserved for data Data (adc mode -> 10, ToT -> 12; see TDR page 161); 4 reserved for gain (currently only 3 gains used; 12 are technically possible and may be used)

Commit:
Merged add_HGCSample_gain_fix_cns_CMSSW_11_0_0_pre10_GF from repository franzoni with cms-merge-topic

pfs committed:
Merged HGC_ageing_SampleBits_forProd_CMSSW_11_0_0_pre11 from repository franzoni with cms-merge-topic

pfs committed:
reshuffling HGCalSiNoiseMap::getSiOpCharacteristics to be easier to handle also in startup scenarios

pfs committed:
factorizing HGCal_setEndOfLifeNoise to allow to be re-used for a realistic startup scenario

pfs committed:
readding arguments for backward compatibility

pfs committed:
debugging customization for realistic startup

Commit:
Merge pull request #10 from PFCal-dev/pre11_realisticstartup

Pre11 realisticstartup

Commit:
Merged HGC_ageing_SampleBits_forProd_CMSSW_11_0_0_pre11 from repository franzoni with cms-merge-topic

Commit:
comments

Commit:
Merge pull request #11 from franzoni/from-CMSSW_11_0_0_pre11_SERVICE

add comments and clarity

Commit:
code checks - replace with cassert

franzoni commented:
please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12630

- This PR adds an extra 224KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12630/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12630/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

Commit:
code-checks

franzoni commented:
please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12632

- This PR adds an extra 120KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.
https://cmssdt.cern.ch/jenkins/job/ib-run-pr-tests/3358/console Started: 2019/11/05 17:33

cmsbuild commented:
A new Pull Request was created by  @franzoni (Giovanni Franzoni) for master.

It involves the following packages:

DataFormats/HGCDigi
SLHCUpgradeSimulations/Configuration
SimCalorimetry/HGCalSimAlgos
SimCalorimetry/HGCalSimProducers


@cmsbuild, @civanch, @kpedro88, @mdhildreth can you please review it and eventually sign? Thanks.
@vandreev11, @sethzenz, @makortel, @sviret, @rovere, @lgray, @cseez, @apsallid, @pfs, @deguio, @ebrondol, @kpedro88 this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-1

Tested at: 92e65f7236037152d597ac5f8f3407ff0a9eb625

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3358/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
20034.0 step3
<pre>runTheMatrix-results/20034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D35+RecoFullGlobal_2026D35+HARVESTFullGlobal_2026D35/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D35+RecoFullGlobal_2026D35+HARVESTFullGlobal_2026D35.log

</pre>20434.0 step3
<pre>runTheMatrix-results/20434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D41_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D41+RecoFullGlobal_2026D41+HARVESTFullGlobal_2026D41/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D41_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D41+RecoFullGlobal_2026D41+HARVESTFullGlobal_2026D41.log

</pre>22034.0 step3
<pre>runTheMatrix-results/22034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D46_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D46+RecoFullGlobal_2026D46+HARVESTFullGlobal_2026D46/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D46_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D46+RecoFullGlobal_2026D46+HARVESTFullGlobal_2026D46.log

</pre>21234.0 step3
<pre>runTheMatrix-results/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D44_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D44+RecoFullGlobal_2026D44+HARVESTFullGlobal_2026D44/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2026D44_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D44+RecoFullGlobal_2026D44+HARVESTFullGlobal_2026D44.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

kpedro88 commented:
```
----- Begin Fatal Exception 05-Nov-2019 18:48:25 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=HGCalLayerClusterProducer label='hgcalLayerClusters'
Exception Message:
Illegal parameter found in configuration.  The parameter is named:
 'scaleByDoseAlgo'
You could be trying to use a parameter name that is not
allowed for this plugin or it could be misspelled.
----- End Fatal Exception -------------------------------------------------
```

kpedro88 commented on DataFormats/HGCDigi/bin/HGCSampleTest.cpp:
```diff
@@ -0,0 +1,107 @@
+  // init static values
+  uint32_t adc = 124;
+  uint32_t gain = 15;
+  uint32_t thrADC = 900;
```
can be const

franzoni replied:
tks, addressed

kpedro88 commented on DataFormats/HGCDigi/bin/HGCSampleTest.cpp:
better to use the more modern [C++ random number library](https://en.cppreference.com/w/cpp/numeric/random) rather than `rand()`

franzoni replied:
tks, addressed

kpedro88 commented on SLHCUpgradeSimulations/Configuration/python/aging.py:
```diff
@@ -59,11 +59,17 @@ def ageHF(process,turnon):
+    from SimCalorimetry.HGCalSimProducers.hgcalDigitizer_cfi import HGCal_setRealisticStartupNoise
+    process = HGCal_setRealisticStartupNoise(process)
+    return process
+
```
remove extra space

franzoni replied:
tks, addressed

kpedro88 commented on SimCalorimetry/HGCalSimAlgos/interface/HGCalSiNoiseMap.h:
```diff
@@ -56,10 +61,10 @@ class HGCalSiNoiseMap : public HGCalRadiationMap {
 
 private:
-  //
+  //vector of three params, per sensor type: 0:120 [mum], 1:200, 2:300
```
if we know these will always have 3 values, could use `std::array` rather than `std::vector`

franzoni replied:
tks, addressed

kpedro88 commented on DataFormats/HGCDigi/interface/HGCSample.h:
```diff
@@ -12,31 +12,48 @@
 public:
-  enum HGCSampleMasks { kThreshMask = 0x1, kModeMask = 0x1, kToAMask = 0x3ff, kDataMask = 0xfff };
-  enum HGCSampleShifts { kThreshShift = 31, kModeShift = 30, kToAShift = 13, kDataShift = 0 };
+  enum HGCSampleMasks {
```
it seems like the changes in this class could break backward compatibility for reading old MC samples.

Currently the class version for `HGCSample` is not tracked, but it probably should be: https://github.com/cms-sw/cmssw/blob/92e65f7236037152d597ac5f8f3407ff0a9eb625/DataFormats/HGCDigi/src/classes_def.xml

franzoni replied:
> class could break backward compatibility for reading old MC samples.

it does indeed. Especially, kToAShift = 13 having changed to kToAShift = 12 (this is allegedly a fix - shifting by 13 was wasting 1 bit) will break backward compatibility.

I am not aware we are aiming at reading in 11_ release, MC productions done with previous cycles. I gather that tracking the version won't solve the incompatibility, it'll rather provide documentation of it ? I can introduce version tracking if requested, I just don't fully know how to do it (I've poked around a bit.. how does one computed the [ckecsksum](https://github.com/cms-sw/cmssw/blob/master/DataFormats/HGCRecHit/src/classes_def.xml#L3)  for instance ?

kpedro88 replied:
In general, people like to be able to read files from the old productions as long as there hasn't been a new production yet (at least if the geometries have stayed the same).

scram can generate the checksum. Here is an example:
```
cmsrel CMSSW_11_0_0_pre11
cd CMSSW_11_0_0_pre11/src
cmsenv
git cms-addpkg Formats/HGCDigi
```
Edit `DataFormats/HGCDigi/src/classes_def.xml` to change `   <class name="HGCSample"/>` to:
```
   <class name="HGCSample" ClassVersion="2">
    <version ClassVersion="2" checksum="0"/>
   </class>
```
then `scram b` will notice the incorrect checksum and provide an error:
```
error: class 'HGCSample' has a different checksum for ClassVersion 3. Increment ClassVersion to 4 and assign it to checksum 3298262857
Suggestion: You can run 'scram build updateclassversion' to generate src/DataFormats/HGCDigi/src/classes_def.xml.generated with updated ClassVersion
```
Running the suggested command `scram b updateclassversion` will generate a new xml file and will show this message:
```
>> Checking src/DataFormats/HGCDigi/src/classes_def.xml for EDM Class Version update
>> New Class Version File:  src/DataFormats/HGCDigi/src/classes_def.xml.generated
```
Then you can just do:
```
mv DataFormats/HGCDigi/src/classes_def.xml.generated DataFormats/HGCDigi/src/classes_def.xml
```
and then delete the dummy version 2 in the xml (recommended to start with version 3).

In this case, you would do this for the version of the class before your changes, then do it again after your changes (which would become version 4).

Further, it is possible to make the change actually backwards compatible by specifying an IO rule (which tells ROOT how to read an old version into the memory structure of the new version when deserializing). You can search the repo for `ioread` to see examples of this, though it is not always straightforward to make it work.

kpedro88 replied:
@franzoni if you need more clarification here, please let me know

franzoni replied:
>  people like to be able to read files from the old productions as long as there hasn't been a new production yet

indeed. Existing large scale production have been done w/ 10_6 (for the L1T TDR). 
The next round will be with 11_0 ( see https://indico.cern.ch/event/856064/contributions/3602726/attachments/1940953/3218266/19-11-07_News_PPD.pdf ) where also geometry will be updated. 

> Here is an example:

Thanks! I've taken also the liberty of pointing this [documentation](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Class_versioning) to this set of instructions. The twiki paragraph may need updating.

>  it is possible to make the change actually backwards compatible by specifying an IO rule

`ioread` is used to handle root serialisation formats, in the examples I've looked at. Here the backward incompatibility is generated by a different interpretation of the bits in the same variable ( `value_` ), which is not resolved by de-serialising root differently. Backward compatibility would need adding a version number to the content of the class or of `value_` itself, and steer the content of `value_` accordingly. In my opinion this would be an overkill at this stage.
Do we concur ?

kpedro88 replied:
The io rule would need to take the input `value_` from a version 3 instance of the product, and then modify it to have the correct bit sequence for version 4. An example of this comes from HcalDetId, which underwent a similar change: https://github.com/cms-sw/cmssw/blob/master/DataFormats/HcalDetId/src/classes_def.xml#L12-L17

But actually, I'm not sure how ROOT will interpret a 10_6 sample (when there was no class version assigned)...

kpedro88 commented:
@franzoni it seems like the changes here shouldn't have a major impact on the CPU usage of the digitization code with aging enabled (compared to previous versions), but has this been explicitly checked?

rovere commented:
From extra quick look, the crash could be related to this line:
https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HGCalRecProducers/interface/HGCalCLUEAlgo.h#L98

Commit:
KP comment: propagate new varialble (scaleByDoseAlgo in hgcalDigitizer_cfi) to HGCalCLUEAlgo; via fillDescriptions of HGCalLayerClusterProducer

Commit:
KP comment:more modern C++ random number library

Commit:
KP comment:remove extra space

Commit:
KP comment:could use std::array rather than std::vector

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12641

- This PR adds an extra 128KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12641/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12641/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

Commit:
uint16 wherever it should; remove saturation check (and ceiling), which was unused also berfore

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12643

- This PR adds an extra 28KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12643/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12643/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

franzoni commented:
thanks @kpedro88 for the detailed review.

> @franzoni it seems like the changes here shouldn't have a major impact on the CPU usage of 
> the digitization code with aging enabled (compared to previous versions), but has this been 
> explicitly checked?

We introduced the `HGCSampleTest.cpp` to keep indeed that under control.
With the commit I've just added, and using `HGCSampleTest.cpp` as a benchmark there's a gain in time, see the **[vanilla pre11]** test versus the **[PR28349]** test below.
The ~10% gain stems from having removed the check wether words being set are larger than the mask of bits available, resulting in ceiling the word to the max value possible; with the current implementation, a word exceeding the max value will just get masked.
This is de facto the same behaviour as was used before this PR (where the ceiling was in one of the method, which was however never used).


**[vanilla pre11]**
```
HGCSampleTest  1000000000
Basic performance tests for HGCSample (pseudo-random seed set according to local time)

num parameters entered: 2
	 + repetitions [int]: 1000000000
	 + generateRandomValues [true/false]: 0


Done 1000000000	1000000000
HGCSampleTest 1000000000  8.55s user 0.04s system 99% cpu 8.640 total

```


**[PR28349]**
```
time HGCSampleTest  1000000000
Basic performance tests for HGCSample (pseudo-random seed set according to local time)

num parameters entered: 2
	 + repetitions [int]: 1000000000
	 + generateRandomValues [true/false]: 0


Done 1000000000	1000000000
HGCSampleTest 1000000000  7.64s user 0.08s system 97% cpu 7.881 total
```

Commit:
code-checks

franzoni commented:
please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12648

- This PR adds an extra 140KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #28349 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 3aa95b2fa8a233fcd21fe694eb9f134c3d78d8c5
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3371/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3371/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 244 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939026
 - DQMHistoTests: Total failures: 497
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2938188
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

slava77 commented on RecoLocalCalo/HGCalRecProducers/interface/HGCalCLUEAlgo.h:
```diff
@@ -97,6 +97,8 @@ class HGCalCLUEAlgo : public HGCalClusteringAlgoBase {
+    descNestedNoiseMIP.add<unsigned int>("scaleByDoseAlgo", 0);
+    iDesc.add<edm::ParameterSetDescription>("scaleByDoseAlgo", descNestedNoiseMIP);
     descNestedNoiseMIP.add<std::string>("doseMap", "");
     iDesc.add<edm::ParameterSetDescription>("doseMap", descNestedNoiseMIP);
```
we already have the following in the config 
```python
    scaleByDose = cms.PSet(
      scaleByDose = cms.bool(False)
    ),
    doseMap = cms.PSet(
      scaleByDose = cms.bool(False),
      doseMap = cms.string('')
    ),
    noiseMip = cms.PSet(
      scaleByDose = cms.bool(False),
      doseMap = cms.string(''),
      noise_MIP = cms.double(0.01)
    ),
```
There will now be also ```scaleByDoseAlgo = cms.PSet(scaleByDose = False, scaleByDoseAlgo = 0)``` with ```scaleByDoseAlgo = 0``` added to all the following PSets.
is this nested accumulating content really desired?

franzoni replied:
thanks @slava77 
I see there are 3 params added to the main pset which are not needed.
I've added a commit which fixes that - at least according to my understanding.
The relvals run.
@deguio (who committed the 3 lines I am removing) @rovere may comment, do you share the logic ?

franzoni commented on DataFormats/HGCDigi/bin/BuildFile.xml:
```diff
@@ -0,0 +1,3 @@
+<use name="DataFormats/HGCDigi" />
+<bin   name="HGCSampleTest" file="HGCSampleTest.cpp">
```
hi @slava77 thanks for your comments.
I gather that moving to /test does not provide an executable which is available in the $PATH.
I found very useful being able to execute at leisure the HGCSampleTest while developing, is there a way to do that ?
Otherwise I'd see good reasons to keep the .cpp in bin.

slava77 replied:
@Dr15Jones @davidlange6 
since this is the first /bin in \*DataFormats, I'm curious if this is accidental or by design
(I didn't find anything against in https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Restrictions_on_the_package_defi

fabiocos replied:
@slava77 IMHO this is just a simple test code as we have many others, and might simply go into a test directory (we have many of them in DataFormats). Discussing in the past with @davidlange6 I think /bin was mostly intended for standalone utilities... Anyway I let others to express themselves

franzoni replied:
hello @fabiocos and all
can one get an executable to be run in parallel to developing, if the .cpp moves to the /test directory ?
Perhaps it was me? ,  when I moved this small program to test, no executable appeared in a folder covered by my $PATH .

franzoni replied:
I also tried with the test programs of one of the DetId directories, and could not execute it.

makortel replied:
> can one get an executable to be run in parallel to developing, if the .cpp moves to the /test directory ?
> Perhaps it was me? , when I moved this small program to test, no executable appeared in a folder covered by my $PATH .

The executables from `test` end up in `$CMSSW_BASE/test/$SCRAM_ARCH` and can be run from there.

Dr15Jones replied:
In addition, executables in `test` can be run via `scram b runtests` and are run in every IB and pull request.

Commit:
slava77 comment: remove unneeded params from main pset

cmsbuild commented:
The code-checks are being triggered in jenkins.

franzoni commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12676

- This PR adds an extra 132KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #28349 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77 can you please check and sign again.

cmsbuild commented:
+1
Tested at: fe44ec559fd7ea5b112f2cb2f0c01e0ddb255366
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3405/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3405/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 244 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2936360
 - DQMHistoTests: Total failures: 526
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2935493
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

slava77 commented on SimCalorimetry/HGCalSimAlgos/src/HGCalRadiationMap.cc:
```diff
@@ -3,7 +3,10 @@
 
 //
-void HGCalRadiationMap::setDoseMap(const std::string& fullpath) { doseMap_ = readDosePars(fullpath); }
+void HGCalRadiationMap::setDoseMap(const std::string& fullpath, const unsigned int& algo) {
```
better pass ```algo``` by value, it costs more to pass by ref in this case

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/28349	fe44ec5 
- reco changes are only in RecoLocalCalo/HGCalRecProducers/interface/HGCalCLUEAlgo.h and look good (including also a fix to some incorrect filling of PSet introduced in the past devels)
- jenkins tests pass and comparisons with the baseline show only (relevant) changes in the phase-2 wf D46 22034.0 apparently related to the digitization changes upstream of reco

Commit:
slava77 comment: better pass algo by value,

franzoni commented:
please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #28349 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 455213684aaa92abbdd70167665dd4b76d712341
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3420/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3420/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 244 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2936360
 - DQMHistoTests: Total failures: 525
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2935494
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

franzoni commented:
hello @kpedro88 @slava77 
all your comments posted so far have been addressed - there's only one item possibly up for discussion form @kpedro88 [here](https://github.com/cms-sw/cmssw/pull/28349#discussion_r342854148), where I've asked a question/advice.

Let us know if there's more info or actions you'd like. Tkx!

slava77 commented:
IIUC, the file in /bin should be moved to /test after the most recent comments

franzoni commented:
Hi @slava77 and all, I'll follow up, ok.

Commit:
slava77 this is the first /bin in DataFormats, better move to /test

Commit:
versioning for HGCSample; version 3 set to the pre11 version (will update in next commit)

Commit:
kpedro88: Currently the class version for HGCSample is not tracked, but it probably should be. Set to V4

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12733

- This PR adds an extra 136KB to repository

cmsbuild commented:
Pull request #28349 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77 can you please check and sign again.

Commit:
kpedro88: Currently the class version for HGCSample is not tracked, but it probably should be. Set to V11 according to https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Class_versioning

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12734

- This PR adds an extra 136KB to repository

cmsbuild commented:
Pull request #28349 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77 can you please check and sign again.

franzoni commented:
> IIUC, the file in /bin should be moved to /test after the most recent comments

@slava77 : done.

franzoni commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 466971147269b3b4a287ba8efe18efa241731592
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3461/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3461/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 244 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2936360
 - DQMHistoTests: Total failures: 497
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2935522
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

slava77 commented on DataFormats/HGCDigi/src/classes_def.xml:
```diff
@@ -1,5 +1,9 @@
-   <class name="HGCSample"/>
+   <class name="HGCSample" ClassVersion="11">
+    <version ClassVersion="11" checksum="2120380856"/>
+    <version ClassVersion="10" checksum="3298262857"/>
```
was the old version really 10?
Was it ever recorded in the persisted data? (IIRC, the default class version defined internally by root is something  either 0 or 1).
IIUC, the "old" version 10 and the checksum for it here have no consequence and can be removed.

New class versions should start from 3
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Class_versioning

kpedro88 replied:
The old version was not recorded. I agree the versioning should start from 3 (had overlooked this before). I thought it was helpful to have the checksum for the old version of the class, but maybe it doesn't matter.

franzoni replied:
The paragraph https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Class_versioning is in many aspects outdated.

Quoting from the twiki: 
>  If a class was previously stored without a class version assigned then one should use 10 as the initial version instead of 3 because ROOT will automatically assign a version number to each unversioned class starting with 3 and increment the version number each time it finds an instance of a class with a different checksum while writing to the file.

we should start from version 10. @kpedro88 had asked I would record the checksum also of the version which is already in CMSSW:
> I thought it was helpful to have the checksum for the old version of the class, but maybe it doesn't matter.

Given this from the twiki:
> .. because ROOT will automatically assign a version number to each unversioned class starting with 3

I'll remove the checksum of the existing dataformat and add the checksum introduced by this PR versioning it to 10.

slava77 reviewed (commented):
DataFormats/HGCDigi/bin/HGCSampleTest.cpp and  DataFormats/HGCDigi/bin/BuildFile.xml are still a part of this PR. They should be removed.

Commit:
slava77 this is the first /bin in DataFormats, better move to /test

cmsbuild commented:
The code-checks are being triggered in jenkins.

franzoni commented:
> They should be removed.
addressed. Thanks @slava77

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12747

- This PR adds an extra 136KB to repository

- Found files with invalid states:
  - DataFormats/HGCDigi/bin/BuildFile.xml:
    - Added: e2fd0221b36037e623944bfb08e9217a8e7b5f59
    - Deleted: 9f2bd2bbac28938a4766fa4a319885957e80ea3b
  - DataFormats/HGCDigi/bin/HGCSampleTest.cpp:
    - Added: e2fd0221b36037e623944bfb08e9217a8e7b5f59
    - Modified: 1669ae5efe2c4663ce012d4ea2708fc0b97019da, b0e3ccbdf2520cee26112389a3d76c92988a183c, 346161f4e78ac5ed2fa6104c43a06654ad564077, 82fb20887e8536237d6534b6e1e5868485ccd992, 92e65f7236037152d597ac5f8f3407ff0a9eb625, cca7a7051dd37b7bb14ae27b768ee762e7a07cb9
    - Deleted: 9f2bd2bbac28938a4766fa4a319885957e80ea3b

cmsbuild commented:
Pull request #28349 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77 can you please check and sign again.

Commit:
kpedro88: Currently the class version for HGCSample is not tracked, but it probably should be. Set to V10

Commit:
kpedro88: Currently the class version for HGCSample is not tracked, but it probably should be. Set to V10

cmsbuild commented:
The code-checks are being triggered in jenkins.

slava77 commented on DataFormats/HGCDigi/src/classes_def.xml:
```diff
@@ -1,5 +1,8 @@
 <lcgdict>
-   <class name="HGCSample"/>
+   <class name="HGCSample" ClassVersion="10">
+    <version ClassVersion="10" checksum="2120380856"/>
```
new class versions should start from 3 https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCreatingNewProducts#Class_versioning

kpedro88 replied:
@slava77 please read https://github.com/cms-sw/cmssw/pull/28349#discussion_r345994831

slava77 replied:
ok, thank you.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12748

- This PR adds an extra 136KB to repository

- Found files with invalid states:
  - DataFormats/HGCDigi/bin/BuildFile.xml:
    - Added: e2fd0221b36037e623944bfb08e9217a8e7b5f59
    - Deleted: 9f2bd2bbac28938a4766fa4a319885957e80ea3b
  - DataFormats/HGCDigi/bin/HGCSampleTest.cpp:
    - Added: e2fd0221b36037e623944bfb08e9217a8e7b5f59
    - Modified: 1669ae5efe2c4663ce012d4ea2708fc0b97019da, b0e3ccbdf2520cee26112389a3d76c92988a183c, 346161f4e78ac5ed2fa6104c43a06654ad564077, 82fb20887e8536237d6534b6e1e5868485ccd992, 92e65f7236037152d597ac5f8f3407ff0a9eb625, cca7a7051dd37b7bb14ae27b768ee762e7a07cb9
    - Deleted: 9f2bd2bbac28938a4766fa4a319885957e80ea3b

cmsbuild commented:
Pull request #28349 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77 can you please check and sign again.

Commit:
first DRAFT of Manual Data Model Evolution

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 70058b86ccdb0e482ffd5e4ee4358b89f676e335
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3476/summary.html

cmsbuild commented:
Comparison job queued.

Commit:
Data Model Evolution: in place (verbose)

Commit:
kpedro88: Data Model Evolution: in place (it is possible to make the change actually backwards compatible by specifying an IO rule)

cmsbuild commented:
The code-checks are being triggered in jenkins.

Commit:
kpedro88: Data Model Evolution: in place (it is possible to make the change actually backwards compatible by specifying an IO rule)

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12763

- This PR adds an extra 148KB to repository

- Found files with invalid states:
  - DataFormats/HGCDigi/bin/BuildFile.xml:
    - Added: e2fd0221b36037e623944bfb08e9217a8e7b5f59
    - Deleted: 9f2bd2bbac28938a4766fa4a319885957e80ea3b
  - DataFormats/HGCDigi/bin/HGCSampleTest.cpp:
    - Added: e2fd0221b36037e623944bfb08e9217a8e7b5f59
    - Modified: 1669ae5efe2c4663ce012d4ea2708fc0b97019da, b0e3ccbdf2520cee26112389a3d76c92988a183c, 346161f4e78ac5ed2fa6104c43a06654ad564077, 82fb20887e8536237d6534b6e1e5868485ccd992, 92e65f7236037152d597ac5f8f3407ff0a9eb625, cca7a7051dd37b7bb14ae27b768ee762e7a07cb9
    - Deleted: 9f2bd2bbac28938a4766fa4a319885957e80ea3b


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12763/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12763/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

franzoni commented:
w/ @pfs we've implemented  schema evolution for `HGCSample`
https://root.cern.ch/root/html/io/DataModelEvolution.html
Thanks @kpedro88 for advice!

With the lats 5 commits, `HGCSample` V9 written with releases up to 11_0_0_pre12 can be read also using the V10 `HGCSample` which is introduced by this PR.

**Validation 1: inspection of a few samples**
We've generated and written out 1 event in 11_0_0_pre11, and printed the `HGCSample` from the HGCalUncalibRecHitRecWeightsAlgo.
Read back the same event in  11_0_0_pre11 + this PR, and printed the `HGCSample` from the HGCalUncalibRecHitRecWeightsAlgo - in this case the `HGCSample` have been de-serialised using the rules defined (by this PR) for the schema evolution. This is the outcome for four which have been inspected:
```

+++ V9  RecWeightsAlgo - id: 2148550960 data: 129 getToAValid: 1 raw: 2149326977
10000000000111000010000010000001
         \--> toa  !\--> data
*** V10 RecWeightsAlgo - id: 2148550960 data: 129 getToAValid: 1 raw: 2685276289
10100000000011100001000010000001
  ~       \--> toa  \-->  data


+++ V9  RecWeightsAlgo - id: 2154385067 data: 341 getToAValid: 1 raw: 2149278037
10000000000110110110000101010101
         \--> toa  !\--> data
*** V10 RecWeightsAlgo - id: 2154385067 data: 341 getToAValid: 1 raw: 2685251925
10100000000011011011000101010101
  ~       \--> toa  \-->  data


+++ V9  RecWeightsAlgo - id: 2154398930 data: 41 getToAValid: 0 raw: 2147483689
10000000000000000000000000101001
         \--> toa  !\--> data
*** V10 RecWeightsAlgo - id: 2154398930 data: 41 getToAValid: 0 raw: 2147483689
10000000000000000000000000101001
  ~       \--> toa  \-->  data


*** V9 RecWeightsAlgo - id: 2154399251 data: 54 getToAValid: 0 raw: 2147483702
10000000000000000000000000110110

+++ V10 RecWeightsAlgo - id: 2154399251 data: 54 getToAValid: 0 raw: 2147483702
10000000000000000000000000110110
```

One can see the toa bits shifting (to remove the 13th bit previously left unused) in V10 w.r.t to V9 , and the boolean bit (from the boolean data member suppressed in V10) appearing at bit 30.

**Validation 2: 10 photon gun events**
10 photon gun events generated in 10_6_x and read form 11_0_0_pre11 [with](https://gitlab.cern.ch/psilva/hgcdigiformatvalidation/blob/master/data/10_6_X_to_11_0_0_pre11_withversioning.diff) and [without](https://gitlab.cern.ch/psilva/hgcdigiformatvalidation/blob/master/data/10_6_X_to_11_0_0_pre11_noversioning.diff) the commits of the schema evolution. Out of 347k `HGCSample` , no differences in toa value and toa valid bit.

franzoni commented:
please test

franzoni commented:
we have addressed all comments received so far.
Please let us know if more actions are needed on our part.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3476/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 244 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2936360
 - DQMHistoTests: Total failures: 497
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2935522
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

Commit:
code-format

cmsbuild commented:
The code-checks are being triggered in jenkins.

franzoni commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12765

- This PR adds an extra 152KB to repository

- Found files with invalid states:
  - DataFormats/HGCDigi/bin/BuildFile.xml:
    - Added: e2fd0221b36037e623944bfb08e9217a8e7b5f59
    - Deleted: 9f2bd2bbac28938a4766fa4a319885957e80ea3b
  - DataFormats/HGCDigi/bin/HGCSampleTest.cpp:
    - Added: e2fd0221b36037e623944bfb08e9217a8e7b5f59
    - Modified: 1669ae5efe2c4663ce012d4ea2708fc0b97019da, b0e3ccbdf2520cee26112389a3d76c92988a183c, 346161f4e78ac5ed2fa6104c43a06654ad564077, 82fb20887e8536237d6534b6e1e5868485ccd992, 92e65f7236037152d597ac5f8f3407ff0a9eb625, cca7a7051dd37b7bb14ae27b768ee762e7a07cb9
    - Deleted: 9f2bd2bbac28938a4766fa4a319885957e80ea3b

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #28349 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 2d60768da0090f003a39dad7a76b7f79e5e3baf8
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3488/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3488/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 244 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2936360
 - DQMHistoTests: Total failures: 497
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2935522
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

kpedro88 commented:
There is a very minor change in the HGCHEB RecHits, which propagates to downstream quantities:
![HGCHEB RecHits](http://kjplanet.com/pr/all_OldVSNew_TTbar14TeV2026D46wf22034p0c_log10HGCRecHitsSorted_HGCalRecHit_HGCHEBRecHits_RECO_obj_obj_energy_11.png)

Is this expected?

slava77 commented on DataFormats/HGCDigi/interface/HGCSample.h:
```diff
@@ -71,6 +71,36 @@ class HGCSample {
+  /**
+     @short Data Model Evolution
+  */
+  static uint32_t convertToNewFormat(uint32_t valueOldForm, bool toaFiredOldForm) {
```
shouldn't this have a version in the name?
"New" will become obsolete after another update.

franzoni replied:
addressed.

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/28349	2d60768 
- updates since the last reco signoff did not affect code in the reco category https://github.com/cms-sw/cmssw/pull/28349/files/fe44ec559fd7ea5b112f2cb2f0c01e0ddb255366..2d60768da0090f003a39dad7a76b7f79e5e3baf8
- jenkins tests pass and comparisons with the baseline are apparently showing the same kind of differences starting from HGCAL hits

franzoni commented:
> Is this expected?

we have an understanding. Working on final checks before we report here.

Commit:
slava77: shouldn't this have a version in the name? convertV9ToV10

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12779

- This PR adds an extra 16KB to repository

- Found files with invalid states:
  - DataFormats/HGCDigi/bin/BuildFile.xml:
    - Added: e2fd0221b36037e623944bfb08e9217a8e7b5f59
    - Deleted: 9f2bd2bbac28938a4766fa4a319885957e80ea3b
  - DataFormats/HGCDigi/bin/HGCSampleTest.cpp:
    - Added: e2fd0221b36037e623944bfb08e9217a8e7b5f59
    - Modified: 1669ae5efe2c4663ce012d4ea2708fc0b97019da, b0e3ccbdf2520cee26112389a3d76c92988a183c, 346161f4e78ac5ed2fa6104c43a06654ad564077, 82fb20887e8536237d6534b6e1e5868485ccd992, 92e65f7236037152d597ac5f8f3407ff0a9eb625, cca7a7051dd37b7bb14ae27b768ee762e7a07cb9
    - Deleted: 9f2bd2bbac28938a4766fa4a319885957e80ea3b

cmsbuild commented:
Pull request #28349 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77 can you please check and sign again.

Commit:
kpedro88: Is this expected? FIX: use 12 bits which are the only ones available

Commit:
kpedro88: Is this expected? esure largest charge convertred in bits is 0xfff==4095, not 4096==0x1000

Commit:
code-format

franzoni commented:
> Is this expected?

It was not expected.

The [2 hits which change amplitude](https://github.com/cms-sw/cmssw/pull/28349#issuecomment-554122677) in the plot (are bit-clamped, hence they both decrease in value) are only in the scintillator section of the HGCAL; they're the result of : 
1) the [feature](https://images.app.goo.gl/S6PxFx4hxgCZV6NT6),  present since pre12, of using  [13 bits in place of the 12 available](https://github.com/cms-sw/cmssw/blob/CMSSW_11_0_0_pre12/SimCalorimetry/HGCalSimProducers/python/hgcalDigitizer_cfi.py#L238) in `HGCSample`
2) having removed the ceiling of the amplitudes in the `HGCSample` itself, to remove the cpu performance penalty ( we discussed this [earlier in this PR](https://github.com/cms-sw/cmssw/pull/28349#issuecomment-550381238l) and given that the HGCAL digitizer is protected against values beyond pow(2, `adcNbits `)
combined.

How @pfs and I propose to solve this, in agreement with @deguio and @rovere (HGCAL DPG) is:

- set the number bits (``adcNbits ``) used for the amplitude of the scintillator section to 12, which is a must
- halve `adcSaturation_fC` (550->275); that won't change number-of-mips/bit currently stored nor the actual saturation value, which is _de factor_ 275 even in the current master (where `HGCSample` implements the ceiling for values above 4095) despite having chosen 550
- ensure that in the scintillator digitizer, whenever the adc value exceeds the 12 bit coverage, the `HGCSample` is set to 4095

Validation: 
We've generated 20 events with relval 21234; 2 hits in the scintillator exceeded the 4095 and got clamped with this PR - we reused these GEN-SIM for incremental tests, in a situation similar to the plot by @kpedro88 .

At first comparison of { 0x DetId , data in adc counts, toavalid } between [2d60768](https://github.com/cms-sw/cmssw/commit/2d60768da0090f003a39dad7a76b7f79e5e3baf8) and IB-CMSSW_11_0_X_2019-11-14-1100.

```
779c779
< 0xa4163e6c 3235 0
---
> 0xa4163e6c 4095 0
831c831
< 0xa418406c 347 0
---
> 0xa418406c 4095 0
```

We found out that the two hits would have 7331 and 4443 counts resp (which are 4096 more than the value . With the three commits which will follow this post, the value 4095 are restored in `HGCSample`.

In a pull request which will follow this one, we'll reassess the usage of bits and choice of least-significant-bit for the scintillator section.

cmsbuild commented:
The code-checks are being triggered in jenkins.

franzoni commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-28349/12792

- This PR adds an extra 152KB to repository

- Found files with invalid states:
  - DataFormats/HGCDigi/bin/BuildFile.xml:
    - Added: e2fd0221b36037e623944bfb08e9217a8e7b5f59
    - Deleted: 9f2bd2bbac28938a4766fa4a319885957e80ea3b
  - DataFormats/HGCDigi/bin/HGCSampleTest.cpp:
    - Added: e2fd0221b36037e623944bfb08e9217a8e7b5f59
    - Modified: 1669ae5efe2c4663ce012d4ea2708fc0b97019da, b0e3ccbdf2520cee26112389a3d76c92988a183c, 346161f4e78ac5ed2fa6104c43a06654ad564077, 82fb20887e8536237d6534b6e1e5868485ccd992, 92e65f7236037152d597ac5f8f3407ff0a9eb625, cca7a7051dd37b7bb14ae27b768ee762e7a07cb9
    - Deleted: 9f2bd2bbac28938a4766fa4a319885957e80ea3b

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #28349 was updated. @perrotta, @civanch, @kpedro88, @cmsbuild, @mdhildreth, @slava77 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 6eb45416dfa871fee41c08b20910d36cc3bf5eb2
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3511/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e191b7/3511/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2936360
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2936018
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

kpedro88 commented:
+upgrade
thanks @franzoni et al. for the careful investigation
this is an opportune time (before the next TDR production) to standardize HGCal digitization practices

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/28349	4c9cb63 
- changes since the last reco signoff for 2d60768 were outside of the reco category
- jenkins tests pass and the comparisons now show no differences

fabiocos commented:
@civanch could you please check and comment or sign it?

@franzoni thanks for addressing the issue with test code in bin directory. The backward compatibility issue has been addressed, so this code looks ready for integration (apart for possible residual comments by simulation)

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
